### PR TITLE
Handle legacy fork id Eth/64

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description
The fork id computation has been updated to not include 0 fork blocks in the list of forks used to compute the fork id.
As a result, previous Besu versions are not capable anymore to talk to newest versions of Besu because their fork id are rejected with the new computation method.
We should allow previous Besu nodes to talk with new Besu nodes by adding a command line flag that enable the legacy fork id computation.
This PR adds a CLI flag to enable the legacy fork id computation method: `--compatibility-eth64-forkid-enabled` defaulted to false.
## Fixed Issue(s)
Adresses #1541

#### Setup
Local testnet with 2 `Besu` nodes:
- `besu/v20.10.1-dev-9589ae08/osx-x86_64/adoptopenjdk-java-11`
    - `--compatibility-eth64-forkid-enabled=true`
- `besu/v1.4.6/osx-x86_64/adoptopenjdk-java-11` 

#### Genesis
```json
{
  "config": {
    "chainId": 1542,
    "homesteadBlock": 0,
    "eip150Block": 0,
    "eip155Block": 0,
    "eip158Block": 4,
    "byzantiumBlock": 5,
    "constantinopleBlock": 6,
    "petersburgBlock": 7,
    "istanbulBlock": 8,
    "muirGlacierBlock": 9,
    "ethash": {
    }
  },
  "difficulty": "0x20",
  "gasLimit": "0x200B20",
  "alloc": {
    "fe3b557e8fb62b89f4916b721be55ceb828dbd73": {
      "privateKey": "8f2a55949038a9610f50fb23b5883af3b4ecb3c3bb792cbcefbd1542c692be63",
      "balance": "90000000000000000000000"
    },
    "627306090abaB3A6e1400e9345bC60c78a8BEf57": {
      "privateKey": "c87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3",
      "balance": "90000000000000000000000"
    },
    "f17f52151EbEF6C7334FAD080c5704D77216b732": {
      "privateKey": "ae6ae8e5ccbfb04590405997ee2d52d2b330726137b875053c36d94e974d162f",
      "balance": "90000000000000000000000"
    }
  }
}
```
#### Result
Nodes are able to communicate correctly using Eth/64 subprotocol.

##### Admin peers of besu/v20.10.1
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": [
        {
            "version": "0x5",
            "name": "besu/v1.4.6/osx-x86_64/adoptopenjdk-java-11",
            "caps": [
                "eth/62",
                "eth/63",
                "eth/64"
            ],
            "network": {
                "localAddress": "127.0.0.1:50958",
                "remoteAddress": "127.0.0.1:40303"
            },
            "port": "0x9d6f",
            "id": "0x3eca270e124b5e24e846bb39d0a911d152cfd2671be079478e72e768363c959852301b40b2afffddbe45a285fd752fa4541e8376f73dad688757e0a07a35e164"
        }
    ]
}
```

##### Admin peers of besu/v1.4.6
```json
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": [
        {
            "version": "0x5",
            "name": "besu/v20.10.1-dev-9589ae08/osx-x86_64/adoptopenjdk-java-11",
            "caps": [
                "eth/62",
                "eth/63",
                "eth/64"
            ],
            "network": {
                "localAddress": "127.0.0.1:40303",
                "remoteAddress": "127.0.0.1:50958"
            },
            "port": "0x765f",
            "id": "0x1c6d296749018e4e4a78baf9a8a3048aae2557c3f2f11a340570d57e71071e2e9816a5f5d9215a333d12b432a81ff5017520b09461c4a102e72c7a1a2d9d7d0f"
        }
    ]
}
```

#### Unit tests

Added `ForkIdBackwardCompatibilityTest` class to cover backward compatibility testing.
Copied legacy `ForkIdManager` from Besu `1.4.6` and added as a test resource to compute expected results for tests.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### CODE REVIEW

**Review by RatanRSur — APPROVED** [review]

**Review by RatanRSur — CHANGES_REQUESTED** [review]
@abdelhamidbakhta discovered some issues so just putting this here to signal that this shouldn't be merged

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java [hunk 1]
```diff
@@ -61,6 +64,9 @@ public ForkIdManager(final Blockchain blockchain, final List<Long> nonFilteredFo
   }
 
   public ForkId computeForkId() {
+    if (legacyEth64) {
+      return forkAndHashList.size() > 0 ? forkAndHashList.get(forkAndHashList.size() - 1) : null;
```
Comment (RatanRSur) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java:68]: If you use `isEmpty` here it's easier to parse at a glance how it differs from the version below

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/LegacyForkIdManager.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    return new ForkId(hash, next);
+  }
+
+  /**
+   * EIP-2124 behaviour
+   *
+   * @param forkId to be validated.
+   * @return boolean (peer valid (true) or invalid (false))
+   */
+  boolean peerCheck(final ForkId forkId) {
```
Comment (RatanRSur) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/LegacyForkIdManager.java:74]: Do we need to keep these unused methods around?

**Review by RatanRSur — COMMENTED** [review]
Which pairs of versions was this tested on?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/ForkIdManager.java]: I agree. Done

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/LegacyForkIdManager.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/LegacyForkIdManager.java]: No. I removed them.

**Comment by abdelhamidbakhta** [comment]
> Which pairs of versions was this tested on?

- 20.10.1 <=> 1.4.6
- 20.10.1 <=> 1.5.0

**Comment by RatanRSur** [comment]
And 1.4.6 and 1.5.0 is expected to continue to not work, correct? We're going to advise upgrading?

**Comment by abdelhamidbakhta** [comment]
> And 1.4.6 and 1.5.0 is expected to continue to not work, correct? We're going to advise upgrading?

Yes this is the plan for the moment.

**Review by RatanRSur — APPROVED** [review]
Ok, cool. Appooved assuming that plan holds.

**Pull request merged**